From 4dbec28a34e5a3d23a433c6511f6575f62c5d73d Mon Sep 17 00:00:00 2001
From: Michal Simek <monstr@monstr.eu>
Date: Fri, 27 May 2011 15:25:09 +0200
Subject: [PATCH] axi_ethernet: Remove loop in BD space checking

Remove do-while loop by checking last BD which
it is necessary for transfer.

Signed-off-by: Michal Simek <monstr@monstr.eu>
---
 drivers/net/ethernet/xilinx/xilinx_axienet_main.c |   20 +++-----------------
 1 files changed, 3 insertions(+), 17 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index 602114e..1cff0c1 100755
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -727,23 +727,9 @@ static inline int axienet_check_tx_bd_space(struct axienet_local *lp,
 								int num_frag)
 {
 	struct axidma_bd *cur_p;
-	int tail;
-
-	tail = lp->tx_bd_tail;
-	cur_p = &lp->tx_bd_v[tail];
-
-	do {
-		if ((cur_p->status) & (XAXIDMA_BD_STS_ACTUAL_LEN_MASK))
-			return NETDEV_TX_BUSY;
-
-		tail++;
-		if (tail >= TX_BD_NUM)
-			tail = 0;
-
-		cur_p = &lp->tx_bd_v[tail];
-		num_frag--;
-	} while (num_frag >= 0);
-
+	cur_p = &lp->tx_bd_v[(lp->tx_bd_tail + num_frag) % TX_BD_NUM];
+	if (cur_p->status & XAXIDMA_BD_STS_ALL_MASK)
+		return NETDEV_TX_BUSY;
 	return 0;
 }
 
-- 
1.7.0.4

